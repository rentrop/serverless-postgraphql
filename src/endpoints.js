
var endpoint = 'http://localhost:5000/graphql';

if(process.env.USE_AWS_ENDPOINT) {
  endpoint = process.env.AWS_ENDPOINT;
};

module.exports = {
    endpoint: endpoint
};
