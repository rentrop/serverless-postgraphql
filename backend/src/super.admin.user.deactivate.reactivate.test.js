import HttpTransport from 'lokka-transport-http';
import Lokka from 'lokka';
import uuidv4 from 'uuid';
import { endpoint } from './endpoints';

const anonLokka = new Lokka({transport: new HttpTransport(endpoint)});
const superAdminEmail = 'superadmin@flo.ods';
const superAdminPassword = 'texasfloods';
const newUserPassword = 'texasfloods';

async function getToken(email, password) {
  const response = await anonLokka.send(`
    mutation($email:String!, $password:String!) {
      authenticate(input: {email: $email, password: $password}) {
        jwtToken
      }
    }
  `,
  {
    email: email,
    password: password
  });

  return response.authenticate.jwtToken;
}

function userInDatabase(userId) {
  it('should see the user in the database', async () => {
    const response = await anonLokka.send(`
      query($userId:Int!){
        userById(id:$userId) {
          firstName
          lastName
          active
        }
      }
    `,
    {
      userId: userId
    });

    expect(response).toMatchSnapshot();
  });
}

describe('When registering, deactivating, and reactivating a user as a super admin', () => {
  var newUserEmail = uuidv4() + '@flo.ods';
  var newUserId;

  describe('As a super admin', async () => {
    var lokka;

    beforeAll(async (done) => {
      getToken(superAdminEmail, superAdminPassword).then((token) => {
        const headers = {
          'Authorization': 'Bearer '+ token
        };
        lokka = new Lokka({transport: new HttpTransport(endpoint, {headers})});
        done();
      });
    });

    it('should register a new user', async () => {
      const response = await lokka.send(`
        mutation($email:String!) {
          registerUser(input: {
            firstName: "New",
            lastName: "User",
            jobTitle: "Community Editor",
            communityId: 1,
            phoneNumber: "555-6666",
            email: $email,
            password:"texasfloods",
            role:"floods_community_editor"
          }) {
            user {
              id
              active
            }
          }
        }
      `,
      {
        email: newUserEmail
      });

      newUserId = response.registerUser.user.id;
      expect(response).not.toBeNull();
      expect(response.registerUser.user.active).toBeTruthy();
    });

    it('after registering', () => {
      userInDatabase(newUserId);
    });
  });

  describe('As the new user', async () => {
    var lokka;

    beforeEach(async (done) => {
      getToken(newUserEmail, newUserPassword).then((token) => {
        const headers = {
          'Authorization': 'Bearer '+ token
        };
        lokka = new Lokka({transport: new HttpTransport(endpoint, {headers})});
        done();
      });
    });

    it('should get the correct current user', async () => {
      const response = await lokka.send(`
        {
          currentUser {
            id
            firstName
          }
        }
      `);

      newUserId = response.currentUser.id;
      expect(response.currentUser.firstName).toMatchSnapshot();
    });
  });

  describe('As a super admin again', async () => {
    var lokka;

    beforeAll(async (done) => {
      getToken(superAdminEmail, superAdminPassword).then((token) => {
        const headers = {
          'Authorization': 'Bearer '+ token
        };
        lokka = new Lokka({transport: new HttpTransport(endpoint, {headers})});
        done();
      });
    });

    it('should deactivate the user', async () => {
      const response = await lokka.send(`
        mutation($userID:Int!) {
          deactivateUser(input: {userId: $userID}) {
            user {
              id
            }
          }
        }
      `,
      {
        userID: newUserId
      });

      expect(response).not.toBeNull();
    });

    it('after deactivating', () => {
      userInDatabase(newUserId);
    });

  });

  describe('As a super admin once more', async () => {
    var lokka;

    beforeAll(async (done) => {
      getToken(superAdminEmail, superAdminPassword).then((token) => {
        const headers = {
          'Authorization': 'Bearer '+ token
        };
        lokka = new Lokka({transport: new HttpTransport(endpoint, {headers})});
        done();
      });
    });

    it('should reactivate the user', async () => {
      const response = await lokka.send(`
        mutation($userId:Int!, $email:String!) {
          reactivateUser(input: {
            userId: $userId,
            email: $email
            password: "texasfloods"
            role: "floods_community_editor"
          }) {
            user {
              firstName
              lastName
              active
            }
          }
        }
      `,
      {
        userId: newUserId,
        email: newUserEmail
      });

      expect(response).toMatchSnapshot();
    });

    it('after reactivating', () => {
      userInDatabase(newUserId);
    });

  });

  describe('As the reactivated user', async () => {
    var lokka;

    beforeAll(async (done) => {
      getToken(newUserEmail, newUserPassword).then((token) => {
        const headers = {
          'Authorization': 'Bearer '+ token
        };
        lokka = new Lokka({transport: new HttpTransport(endpoint, {headers})});
        done();
      });
    });

    it('should get the correct current user', async () => {
      const response = await lokka.send(`
        {
          currentUser {
            firstName
            lastName
          }
        }
      `);

      expect(response).toMatchSnapshot();
    });
  });
});
