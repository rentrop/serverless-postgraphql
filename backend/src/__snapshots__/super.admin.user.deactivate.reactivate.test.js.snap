// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When registering, deactivating, and reactivating a user as a super admin As a super admin once more should reactivate the user 1`] = `
Object {
  "reactivateUser": Object {
    "user": Object {
      "active": true,
      "firstName": "New",
      "lastName": "User",
    },
  },
}
`;

exports[`When registering, deactivating, and reactivating a user as a super admin As the new user should get the correct current user 1`] = `"New"`;

exports[`When registering, deactivating, and reactivating a user as a super admin As the reactivated user should get the correct current user 1`] = `
Object {
  "currentUser": Object {
    "firstName": "New",
    "lastName": "User",
  },
}
`;
